jubatus/core/clustering/util.cpp

```diff
@@ -176,13 +189,47 @@ double dist(const common::sfv_t& p1, const common::sfv_t& p2) {
   return std::sqrt(ret);
 }
 
-double dist(const weighted_point &d1, const weighted_point &d2) {
-  return dist(d1.data, d2.data);
+double point_euclid_dist(const weighted_point &d1, const weighted_point &d2) {
+  return sfv_euclid_dist(d1.data, d2.data);
+}
+
+double sfv_cosine_dist(const common::sfv_t& p1, const common::sfv_t& p2) {
+  double p1_norm = calc_l2norm(p1);
+  double p2_norm = calc_l2norm(p2);
+  if (p1_norm == 0.f || p2_norm == 0.f) {
+    return 1.f;
+  }
+
+  size_t i1 = 0;
+  size_t i2 = 0;
+  double ret = 0.;
+  while (i1 < p1.size() && i2 < p2.size()) {
```

Comment:
It seems necessary to sort ``p1`` and ``p2`` by key. Is not it necessary?

Reply:
Yes, you're right. I think sorting vectors here is better.
I thought all `sfv_vector` in clustering were sorted before calling distance methods. But actually, the sort is done in the driver like [this](https://github.com/jubatus/jubatus_core/blob/master/jubatus/core/driver/clustering.cpp#L132). Therefore if a user uses clustering directly, without the driver, this code may not work correctly.

Reply:
I have found that sorting in distance method causes many deletion of `const` qualifier.
And `embedded_jubatus` work with current codes because it uses `driver`.
For those reasons, I think current implementation is good enough.